config/default.yml

```diff
@@ -2730,6 +2730,7 @@ Style/FrozenStringLiteralComment:
     # `never` will enforce that the frozen string literal comment does not
     # exist in a file.
     - never
+  SafeAutoCorrect: false
```

Comment:
I think the cop is actually unsafe (as it doesn't actually check the source to figure out if it's ok to add the magic comment). Unsafe for a cop means that it yields false positives or the suggestions can change the meaning of the code. Unsafe for auto-correct means the cop the cop suggestions are safe, but the auto-correct can't apply those reliably.

Reply:
I get it! That makes sense to me. Thank for the explanation.

Reply:
> I think the cop is actually unsafe

Makes sense. Should we add `Safe: false` then as well?

Having looked into how `--safe` and `--safe-auto-correct` work I am a bit confused, though:

Is it correct that currently `rubocop --safe-auto-correct` applies "safe" autocorrections of unsafe cops (i.e. it also fixes false positives reliably)? And that you need to use `rubocop --safe-auto-correct --safe` to really not make any changes to the meaning of the code? 😕

Reply:
> Is it correct that currently rubocop --safe-auto-correct applies "safe" autocorrections of unsafe cops (i.e. it also fixes false positives reliably)?

There are no such cases - if a cop is unsafe, its auto-correct is unsafe as well. 

> And that you need to use rubocop --safe-auto-correct --safe to really not make any changes to the meaning of the code?

Yep - basically you get the safe report, but not all of those offences would have safe fixes, so you need to specify this further. I guess that's somewhat confusing, but we couldn't come up with a better way to express it.

Reply:
> Makes sense. Should we add Safe: false then as well?

Only it - this will affect the auto-correct as well.

Reply:
> Only it - this will affect the auto-correct as well.

Are you sure? I tried something along these lines this morning with a cop that only has `Safe: false`:
https://github.com/rubocop-hq/rubocop/blob/8d071f0c9a32f594b696cb7a3b6a0e5adabad42b/config/default.yml#L3344-L3354
```sh-session
$ cat example.rb          
# frozen_string_literal: true

# thief is not a Hash
thief.has_key?(:foo)

$ rubocop -f c example.rb 
example.rb:4:7: C: Style/PreferredHashMethods: Use Hash#key? instead of Hash#has_key?.
thief.has_key?(:foo)
      ^^^^^^^^

1 file inspected, 1 offense detected

$ rubocop -f c --safe example.rb

1 file inspected, no offenses detected

$ rubocop -f c --safe-auto-correct example.rb
example.rb:4:7: C: [Corrected] Style/PreferredHashMethods: Use Hash#key? instead of Hash#has_key?.
thief.has_key?(:foo)
      ^^^^^^^^

1 file inspected, 1 offense detected, 1 offense corrected
```
Here the (wrong) autocorrect is done although the cop is safe.

Reply:
`PreferredHashMethod` is inherently unsafe, as you can never be certain if the receiver is a hash or not. I guess it's simply not marked as unsafe.

Reply:
That's my point. The cop _is_ marked as unsafe, but its auto-correction is not marked as unsafe. You claimed above that setting `Safe: false` for a cop implies `SafeAutoCorrect: false` (if I understood you correctly). My example shows that this is not true.

So we either have a bug here or I still have a misconception how these two levels of safeness are supposed to work.

Reply:
Or I do. 😆 Since I don't spend as much time on the code as I used to, I'm starting to forget some of my original reasoning. But in general I'm struggling to see how the cop's suggestions can be unsafe, but it corrections can be somehow safe. Safe always meant preserving the semantics on the code intact.

Reply:
> I think the cop is actually unsafe (as it doesn't actually check the source to figure out if it's ok to add the magic comment). Unsafe for a cop means that it yields false positives or the suggestions can change the meaning of the code. Unsafe for auto-correct means the cop the cop suggestions are safe, but the auto-correct can't apply those reliably.

I would propose a different definition.  Unsafe for a cop means that it yields false positives *and that's it*. (e.g. "don't use `dig` with single argument is unsafe, as it could be a custom method in a gem dependency and thus perfectly o.k. and without another alternative). The FrozenStringLiteral cop is a good example where the fact that we can't reliably autocorrect a source doesn't mean it can't or shouldn't be done. Since *any source* can modified to have a frozen string literal comment, that all sources should, and that the cop will always correctly detect if there is a frozen string literal comment, I believe this cop is Safe, only its auto-correction is unsafe. How does that sound @bbatsov ?

Reply:
ping @bbatsov